pkg/kv/store_message.go

```diff
@@ -149,26 +146,45 @@ func (s *SecondaryIterator) Next() bool {
 	return s.itr.Next()
 }
 
-func (s *SecondaryIterator) Entry(key *string, value *protoreflect.ProtoMessage) error {
+func (s *SecondaryIterator) Entry() *MessageEntry {
 	secondary := s.itr.Entry()
 	if secondary == nil {
-		return ErrNotFound
-	}
-
-	if value != nil {
-		data, err := s.store.Get(s.ctx, secondary.Value)
-		if err != nil {
-			return fmt.Errorf("getting value from key (secondary key %s, key %s): %w", secondary.Key, secondary.Value, err)
+		return nil
+	}
+	next := secondary.Value.(*SecondaryIndex)
+
+	var (
+		primary *ValueWithPredicate
+		err     error
+	)
+	for {
+		primary, err = s.store.Get(s.ctx, next.PrimaryKey)
+		if !errors.Is(err, ErrNotFound) {
+			break
 		}
-		err = proto.Unmarshal(data.Value, *value)
-		if err != nil {
-			return fmt.Errorf("unmarshal proto data for key %s: %w", string(secondary.Key), err)
+		if !s.itr.Next() {
+			return nil
 		}
+		secondary = s.itr.Entry()
+		if secondary == nil {
+			return nil
+		}
+		next = secondary.Value.(*SecondaryIndex)
+	}
+	if err != nil {
+		s.err = fmt.Errorf("getting value from key (primary key %s): %w", next.PrimaryKey, err)
+		return nil
+	}
```

Comment:
I understand what you're trying to do here, but I think there's a bug since for a successful `Get` from the `store`, you would continue to loop when you should exit the for loop.

Reply:
And like my previous comment, this logic should move to `next`.

Reply:
Oh right, sneaky you